<a href="https://colab.research.google.com/github/keethu12345/TDS/blob/main/Week_2_Graded_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Answers are the ones in **BOLD**.

Question 1:

Which of these Google Sheets functions can process CSV files?


a) IMPORTDATA , IMPORTRANGE

**b) IMPORTDATA, IMPORTXML**

c) IMPORTXML, IMPORTFEED

d) IMPORTRANGE ,IMPORTDATA


Questio 2:

Your objective is to scrape share prices of a portfolio of companies using Excel. You've been successful in loading the share prices along with other information available in the table from the webpage. But as per client request you are required to add another column to the table that labels the stock as "buy" when the % change in the share price exceeds a certain value. Choose the most appropriate option.


a) An additional column cannot be added to the data loaded from the website using Excel.

**b) We cannot insert a column inside an Excel connection range but can add a new column "% change" outside it in Excel.**

c) An additional column can be added but the specific client request cannot be met using Excel. We need Python.

d) We can insert a column between the columns of the connection range in Excel to calculate the "% change".

Question 3:


Using PokeAPI, find out how many moves sandshrew
has that vulpix  does not in in the ultra-sun-ultra-moon version.

In [1]:
#importing required library
import requests

#making an api call to PokeAPI to get the details about the pokemon
response_rapidash = requests.get('https://pokeapi.co/api/v2/pokemon/sandshrew')
response_voltorb = requests.get('https://pokeapi.co/api/v2/pokemon/vulpix')

#parsing through the json response to get the required moves in a particular version
version = 'ultra-sun-ultra-moon'

voltorb_moves = []
for i in response_voltorb.json()['moves']:
  flag = False
  for j in i['version_group_details']:
    if j['version_group']['name'] == version:
      flag = True
      break
  if flag:
    voltorb_moves.append(i['move']['name'])
# print(voltorb_moves)

rapidash_moves = []
for i in response_rapidash.json()['moves']:
  flag = False
  for j in i['version_group_details']:
    # print(j['move_learn_method'])
    if j['version_group']['name'] == version:
      flag = True
      break
  if flag:
    # print(i)
    rapidash_moves.append(i['move']['name'])
# print(rapidash_moves)

#counting the moves that rapidash has but voltorb does not
count = 0
for i in rapidash_moves:
  if i not in voltorb_moves:
    count = count+1
count

43

Question 4:


How many elements match the CSS selector [class^="menu-"], [class*=" menu-"] on this White House page snapshot? (Note the selector carefully Ignore JavaScript-generated elements.)

Answer: 7

1. Inspect element Tab
2. In console write:

`document.querySelectorAll('[class^="menu-"], [class*=" menu-"]').length`


Question 5:

What is the minimum latitude of the bounding box of the city Lagos
in the country Nigeria on the Nominatim API? Pick the osm_id ending with 5124.

Answer: 6.2950575

In [4]:
addr = geolocator.geocode(location_name)
addr.raw # check your osm_id as per qn

{'place_id': 245673267,
 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. http://osm.org/copyright',
 'osm_type': 'node',
 'osm_id': 1886594378,
 'lat': '31.5656822',
 'lon': '74.3141829',
 'class': 'place',
 'type': 'city',
 'place_rank': 16,
 'importance': 0.6036041069355744,
 'addresstype': 'city',
 'name': 'لاہور',
 'display_name': 'لاہور, ضلع لاہور, لاہور ڈویژن, پنجاب, 54500, پاکستان',
 'boundingbox': ['31.4056822', '31.7256822', '74.1541829', '74.4741829']}

In [2]:

from geopy.geocoders import Nominatim
import geopy #.geocoders.options

# Set your custom user-agent
custom_user_agent = "my-custom-application"

# Override the default user_agent
geopy.geocoders.options.default_user_agent = custom_user_agent

# Create a Nominatim geocoder with the custom user-agent
geolocator = Nominatim(user_agent=custom_user_agent)

def get_bounding_box_max_latitude(location_name):
    location = geolocator.geocode(location_name)
    if location:
        bbox = location.raw['boundingbox']
        max_latitude = float(bbox[1])
        return max_latitude
    else:
        return None

# Example usage for Mexico City, Mexico
location_name = "Lagos, Nigeria" # change as per given
max_latitude = get_bounding_box_max_latitude(location_name)
if max_latitude:
    print(f"The maximum latitude of the bounding box for {location_name} is {max_latitude}")
else:
    print(f"Location not found or an error occurred.")

The maximum latitude of the bounding box for Lagos, Nigeria is 6.6150575


Question 6:

What is the maximum latitude of the bounding box of the city Lahore
in the country Pakistan on the Nominatim API? Pick the osm_id ending with 4378.

Answer:  31.7256822

In [3]:

from geopy.geocoders import Nominatim
import geopy #.geocoders.options

# Set your custom user-agent
custom_user_agent = "my-custom-application"

# Override the default user_agent
geopy.geocoders.options.default_user_agent = custom_user_agent

# Create a Nominatim geocoder with the custom user-agent
geolocator = Nominatim(user_agent=custom_user_agent)

def get_bounding_box_max_latitude(location_name):
    location = geolocator.geocode(location_name)
    if location:
        bbox = location.raw['boundingbox']
        max_latitude = float(bbox[1])
        return max_latitude
    else:
        return None

# Example usage for Mexico City, Mexico
location_name = "Lahore, Pakistan" # change as per given
max_latitude = get_bounding_box_max_latitude(location_name)
if max_latitude:
    print(f"The maximum latitude of the bounding box for {location_name} is {max_latitude}")
else:
    print(f"Location not found or an error occurred.")

The maximum latitude of the bounding box for Lahore, Pakistan is 31.7256822


Question 7:

Which of these responses from a requests.get(url)
returns the raw response as bytes (not a string)?

a) response.summary

b) response.text

**c) response.content**

d) response.json

Question 8:

You have the following HTML structure representing a list of featured products on an e-commerce site. Which CSS selector would you use to extract the product names, i.e. text content of only the _span_ elements that are direct children of _div_  elements with the class "highlighted-product" ?



In [ ]:
'''
<div class="highlighted-product">
    <span>Product 1</span>
    <p><span>Special Offer</span></p>
</div>
<div class="highlighted-product">
    <span>Product 2</span>
</div>
<div>
    <span>Product 3</span>
</div>
'''

Answer: div.highlighted-product > span

Question 9:



Given the following nested HTML structure representing a navigation menu, which CSS selector would extract only the _li_ elements that are descendants of the _ul_ with the class "main-menu" and are also the last child within their parent _ul_.

In [ ]:
<ul class="main-menu">
    <li>Home</li>
    <li>About Us</li>
    <ul>
        <li>Our Team</li>
        <li>Our History</li>
        <li>Contact Us</li>
    </ul>
    <li>Services</li>
</ul>
<ul class="main-menu">
    <li>Products</li>
    <li>Support</li>
    <ul>
        <li>FAQ</li>
        <li>Customer Service</li>
    </ul>
</ul>

Answer:  .main-menu li:last-child

In [5]:
!pip install tabula-py

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 32.0 MB/s eta 0:00:00


Question 10:


This PDF file has the marks of students in 5 subjects. Each page has the marks of students in a group. Each row has the marks of a single student.

What is the total  Biology marks of students who scored more 78 or more marks in Economics in groups 1-36 (including both groups)?

In [ ]:
import tabula
import pandas as pd
import numpy as np

l = tabula.read_pdf('/content/ga2.9.pdf', pages = "1-36") #change the "pages" as per your question
l = np.array(l)
l = l.reshape(-1,5)

df = pd.DataFrame(data=l)
df.columns = ['Maths', 'Physics', 'English' ,'Economics' ,'Biology']

df[df["Economics"] >= 78].sum() #change the condition as per your question
